samza-sql-avro/src/main/java/org/apache/samza/sql/avro/AvroTypeFactoryImpl.java

```diff
@@ -0,0 +1,133 @@
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+*
+*   http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+
+package org.apache.samza.sql.avro;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.avro.Schema;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rel.type.RelDataTypeFieldImpl;
+import org.apache.calcite.rel.type.RelDataTypeSystem;
+import org.apache.calcite.rel.type.RelRecordType;
+import org.apache.calcite.sql.type.ArraySqlType;
+import org.apache.calcite.sql.type.SqlTypeFactoryImpl;
+import org.apache.calcite.sql.type.SqlTypeName;
+import org.apache.samza.SamzaException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * TypeFactory that creates the Calcite relational types from the Avro Schema. This is used by the
+ * AvroRelToSamzaMsgConverter to convert the Avro schema to calcite relational schema.
+ */
+public class AvroTypeFactoryImpl extends SqlTypeFactoryImpl {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AvroTypeFactoryImpl.class);
+
+  public AvroTypeFactoryImpl() {
+    super(RelDataTypeSystem.DEFAULT);
+  }
+
+  public RelDataType createType(Schema schema) {
+    Schema.Type type = schema.getType();
+    if (type != Schema.Type.RECORD) {
+      String msg =
+          String.format("System supports only RECORD as top level avro type, But the Schema's type is %s", type);
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    List<RelDataTypeField> relFields = getRelFields(schema.getFields());
+    return new RelRecordType(relFields);
+  }
+
+  private List<RelDataTypeField> getRelFields(List<Schema.Field> fields) {
+    List<RelDataTypeField> relFields = new ArrayList<>();
+    relFields.add(new RelDataTypeFieldImpl(AvroSamzaToRelMsgConverterFactory.KEY_NAME, 0,
+        createTypeWithNullability(createSqlType(SqlTypeName.VARCHAR), true)));
+
+    for (Schema.Field field : fields) {
+      String fieldName = field.name();
+      int fieldPos = field.pos() + 1;
+      RelDataType dataType = getRelDataType(field.schema());
+      relFields.add(new RelDataTypeFieldImpl(fieldName, fieldPos, dataType));
+    }
+
+    return relFields;
+  }
+
+  private RelDataType getRelDataType(Schema fieldSchema) {
+    switch (fieldSchema.getType()) {
+      case ARRAY:
+        // JavaTypeFactoryImpl converts Array into Array(ANY, ANY)
+        return new ArraySqlType(createSqlType(SqlTypeName.ANY), true);
```

Comment:
Yup we do lose this type in this case, when we convert from Avro to Rel. Since these are custom objects, We don't have any other option. I will document this.